File: web/src/types/payment.ts
// Payment types - mirrors internal/domain/payment.go

import type { Branch } from './branch'
import type { Customer } from './customer'
import type { Loan } from './loan'

export type PaymentMethod = 'cash' | 'card' | 'transfer' | 'check' | 'other'
export type PaymentStatus = 'completed' | 'pending' | 'reversed' | 'failed'

export const PAYMENT_METHODS: { value: PaymentMethod; label: string }[] = [
  { value: 'cash', label: 'Efectivo' },
  { value: 'card', label: 'Tarjeta' },
  { value: 'transfer', label: 'Transferencia' },
  { value: 'check', label: 'Cheque' },
  { value: 'other', label: 'Otro' },
]

export const PAYMENT_STATUSES: { value: PaymentStatus; label: string; color: string }[] = [
  { value: 'completed', label: 'Completado', color: 'green' },
  { value: 'pending', label: 'Pendiente', color: 'yellow' },
  { value: 'reversed', label: 'Revertido', color: 'red' },
  { value: 'failed', label: 'Fallido', color: 'gray' },
]

export interface Payment {
  id: number
  payment_number: string
  branch_id: number
  loan_id: number
  customer_id: number

  // Payment details
  amount: number
  principal_amount: number
  interest_amount: number
  late_fee_amount: number

  // Method
  payment_method: PaymentMethod
  reference_number?: string

  // Status
  status: PaymentStatus
  payment_date: string

  // Balances after payment
  loan_balance_after: number
  interest_balance_after: number

  // Reversal info
  reversed_at?: string
  reversed_by?: number
  reversal_reason?: string

  // Notes
  notes?: string

  // Cash session reference
  cash_session_id?: number

  // Audit
  created_by?: number
  created_at: string
  updated_at: string

  // Relations
  branch?: Branch
  loan?: Loan
  customer?: Customer
}

export interface CreatePaymentInput {
  loan_id: number
  amount: number
  payment_method: PaymentMethod
  reference_number?: string
  notes?: string
}

export interface ReversePaymentInput {
  reason: string
}

export interface PaymentListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  search?: string
  branch_id?: number
  loan_id?: number
  customer_id?: number
  status?: PaymentStatus
  payment_method?: PaymentMethod
  date_from?: string
  date_to?: string
}

export interface PayoffCalculation {
  loan_id: number
  principal_remaining: number
  interest_remaining: number
  late_fee_amount: number
  total_payoff: number
}

export interface MinimumPaymentCalculation {
  loan_id: number
  minimum_amount: number
  due_date: string
  installment_number?: number
}

File: internal/domain/payment.go
package domain

import (
	"time"
)

// PaymentMethod represents the method of payment
type PaymentMethod string

const (
	PaymentMethodCash     PaymentMethod = "cash"
	PaymentMethodCard     PaymentMethod = "card"
	PaymentMethodTransfer PaymentMethod = "transfer"
	PaymentMethodCheck    PaymentMethod = "check"
	PaymentMethodOther    PaymentMethod = "other"
)

// PaymentStatus represents the status of a payment
type PaymentStatus string

const (
	PaymentStatusCompleted PaymentStatus = "completed"
	PaymentStatusPending   PaymentStatus = "pending"
	PaymentStatusReversed  PaymentStatus = "reversed"
	PaymentStatusFailed    PaymentStatus = "failed"
)

// Payment represents a loan payment
type Payment struct {
	ID            int64  `json:"id"`
	PaymentNumber string `json:"payment_number"`
	BranchID      int64  `json:"branch_id"`
	LoanID        int64  `json:"loan_id"`
	CustomerID    int64  `json:"customer_id"`

	// Payment details
	Amount          float64 `json:"amount"`
	PrincipalAmount float64 `json:"principal_amount"`
	InterestAmount  float64 `json:"interest_amount"`
	LateFeeAmount   float64 `json:"late_fee_amount"`

	// Method
	PaymentMethod   PaymentMethod `json:"payment_method"`
	ReferenceNumber string        `json:"reference_number,omitempty"`

	// Status
	Status      PaymentStatus `json:"status"`
	PaymentDate time.Time     `json:"payment_date"`

	// Balances after payment
	LoanBalanceAfter     float64 `json:"loan_balance_after"`
	InterestBalanceAfter float64 `json:"interest_balance_after"`

	// Reversal info
	ReversedAt      *time.Time `json:"reversed_at,omitempty"`
	ReversedBy      *int64     `json:"reversed_by,omitempty"`
	ReversalReason  string     `json:"reversal_reason,omitempty"`

	// Notes
	Notes string `json:"notes,omitempty"`

	// Cash session reference
	CashSessionID *int64 `json:"cash_session_id,omitempty"`

	// Audit
	CreatedBy int64 `json:"created_by,omitempty"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Relations
	Branch   *Branch   `json:"branch,omitempty"`
	Loan     *Loan     `json:"loan,omitempty"`
	Customer *Customer `json:"customer,omitempty"`
}

// TableName returns the database table name
func (Payment) TableName() string {
	return "payments"
}

// IsReversed checks if the payment has been reversed
func (p *Payment) IsReversed() bool {
	return p.Status == PaymentStatusReversed
}

// CanBeReversed checks if the payment can be reversed
func (p *Payment) CanBeReversed() bool {
	return p.Status == PaymentStatusCompleted
}
